relay-server/src/actors/global_config.rs

```diff
@@ -0,0 +1,153 @@
+    type Response = AsyncResponse<watch::Receiver<Arc<GlobalConfig>>>;
+
+    fn from_message(_: Subscribe, sender: Sender<watch::Receiver<Arc<GlobalConfig>>>) -> Self {
+        Self::Subscribe(sender)
+    }
+}
+
+impl GlobalConfigService {
+    /// Creates a new [`GlobalConfigService`].
+    pub fn new(upstream: Addr<UpstreamRelay>) -> Self {
+        let (sender, _) = watch::channel(Arc::new(GlobalConfig::default()));
+        Self {
+            sender: Arc::new(sender),
+            upstream,
+            fetch_interval: 10,
+        }
+    }
+
+    fn handle_message(&self, message: GlobalConfigInterface) {
+        match message {
+            GlobalConfigInterface::Get(sender) => {
+                sender.send(self.sender.borrow().clone());
+            }
+            GlobalConfigInterface::Subscribe(sender) => {
+                sender.send(self.sender.subscribe());
+            }
+        }
+    }
+
+    /// Requests a new global config from upstream.
+    ///
+    /// This function does not check any authentication since the upstream
+    /// service deals with it (including observability):
+    /// - If Relay is authenticated, requests will complete as expected.
+    /// - If Relay isn't authenticated yet, the following request to successfull
+    /// authentication will happen.
+    /// - If Relay is permanently blocked, the upstream service will error on
+    /// requests.
+    fn update_global_config(&self) {
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+        let sender = Arc::clone(&self.sender);
+
+        tokio::spawn(async move {
+            let query = GetProjectStates {
+                public_keys: vec![],
+                full_config: false,
+                no_cache: false,
+                global: true,
+            };
+
+            match upstream_relay.send(SendQuery(query)).await {
+                Ok(res) => Self::handle_upstream_response(res, sender),
+                Err(_) => relay_log::error!(
+                    "GlobalConfigService failed to send request to UpstreamService"
+                ),
+            };
+        });
+    }
+
+    fn handle_upstream_response(
+        response: Result<GetProjectStatesResponse, UpstreamRequestError>,
+        sender: Arc<watch::Sender<Arc<GlobalConfig>>>,
+    ) {
+        if let Ok(config) = response {
+            match config.global {
+                Some(global_config) => _ = sender.send(Arc::new(global_config)),
+                None => relay_log::error!("Global config missing in upstream response"),
+            };
+        }
+    }
+}
+
+impl Service for GlobalConfigService {
+    type Interface = GlobalConfigInterface;
+
+    fn spawn_handler(self, mut rx: relay_system::Receiver<Self::Interface>) {
+        tokio::spawn(async move {
+            relay_log::info!("global config service starting");
+
+            let mut ticker = tokio::time::interval(Duration::from_secs(self.fetch_interval));
```

Comment:
This is racy, since the interval does not wait for the upstream request to succeed. If the upstream request takes longer than the interval, multiple concurrent requests will be scheduled. 

Please look at services like the `OutcomeAggregator` or `UpstreamProjectSource`. They use a `SleepHandle` to solve this situation. Here's an adapted implementation to get you started:

 1. Initialize an idle handle in the constructor
 2. Fetch immediately when spawning the service
 3. After fetching, call `schedule_fetch`

```rust
struct GlobalConfigService {
    // ...
    fetch_interval: Duration,
    fetch_handle: SleepHandle, // Initialize with ::idle() in constructor
}

impl GlobalConfigService {
    /// Schedules the next trigger for fetching the global config.
    fn schedule_fetch(&mut self) {
        if self.fetch_handle.is_idle() {
            self.fetch_handle.set(self.fetch_interval);
        }
    }
}

impl Service for GlobalConfigService {
    fn spawn_handler(mut self, mut rx: relay_system::Receiver<Self::Interface>) {
        tokio::spawn(async move {
            self.do_fetch();

            loop {
                tokio::select! {
                    biased;
                    () = &mut self.fetch_handle => self.do_fetch(),
                    // ...
                }
            }
        });
    }
}
```